Feature/create a component detector for ruby 68

PuZZleDucK opened the pull request:
<!--- Tip: You don't have to remove these comments -->

## Description
<!--- Describe your changes in detail -->
Add a component detector and associated package inspector for Ruby.

There were a few tests (around PrettierUsedPractice) that filed on my local machine, but since I didn't change anything there, hopefully that's just me.

I've made a few assumptions in this PR regarding ambiguous versioning. I've taken the simplest route for now and pick the first version specified before a comment and disregard the specifier type (like ~> and >=). Not sure if this is right, but happy to update if there is a way to include the specifiers and ranges. I've got a couple of more specific edge-case questions on the issue here: #68 but I figured a simple solution might do for version 1.

I am also not filtering the dependency type (Runtime/Dev/...) as ruby has a different taxonomy and adding it seemed like a bigger change (although it wouldn't be impossible to add Dev). There are also a few ways to specify the type so it would significantly complicate the package parser too. Happy to look at adding that if you think we need it too.

I've also made the assumption that if a version number isn't known I've made the version number 0.0.0... maybe there is a better way to specify 'don't know'? or maybe version should be removed all together?

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Resolves #68 (Create a component detector for Ruby)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] n/a - I have updated the documentation accordingly.
- [ ] n/a - I have added new practice to practice list in README.md.
- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

adelkahomolova commented:
Hi, @PuZZleDucK I answered your question in #68 . Let me know if it corresponds with your PR. I'll review it after that. :)

PuZZleDucK commented:
I've updated the PR with a couple of changes as suggested. I couldn't work out how to save a package without a version (looks like a PackageVersion is required by the Package interface) so it's only recording the known versions. It will pick the greatest version available, and I've added a test for that too.

It feels a bit awkward coercing the PackageVersion back to a semver for comparison... maybe we should add a 'gte' (greater than or equal to) function for PackageVersions (but I couldn't work out how to do this for an interface... so you'll have to give me a hint if you want me to tackle that one, maybe it should be defined in PackageInspectorBase? or maybe PackageVersions should be upgraded to a class? 🤔)

I think the failing test in CI is intermittent (at least it is on my local machine) but I couldn't find a place to re-trigger it 🤞.

adelkahomolova commented:
Hi @PuZZleDucK , thank you for your changes! The tests are failing just because of a small GitHub API rate limit. Don't worry; it wouldn't affect the merging of your PR :) Only people in the core team can re-trigger it. So I'll do it.

To your question. I really appreciate your PR and the effort to keep the very high quality of code. I don't see an easy solution for this, so I'd be fine with it, although I think it deserves a better solution either. I'll discuss it with my colleague so that we can find the best approach all together.

adelkahomolova commented:
Hi, @PuZZleDucK , so I discuss it witch my colleagues and your solution is good. I'll just ask you for adding this comment `TODO - detect lockfileVersion from lockfile as lockfileVersion doesn't have to be the same as the requiredVersion`. I'll approve it after this small add and merge it. Thank you!

adelkahomolova commented:
Hi, @PuZZleDucK , one more request for you. Can you add the ✅ in the readme in support languages? And also the comment I asked you for in above comment. Thank you!

adelkahomolova commented:
Hi, @PuZZleDucK , did you get my comments? :)

adelkahomolova commented on src/inspectors/package/RubyPackageInspector.ts:
```diff
@@ -0,0 +1,99 @@
+          dependencyType: depType,
+          name: dependency.packageName,
+          requestedVersion: dependency.version,
+          lockfileVersion: dependency.version,
```
// TODO - detect lockfileVersion from lockfile as lockfileVersion doesn't have to be the same as the requiredVersion

adelkahomolova replied:
```suggestion
		// TODO - detect lockfileVersion from lockfile as lockfileVersion doesn't have to be the same as the requested version
          lockfileVersion: dependency.version,
```

prokopsimek commented:
:tada: This PR is included in version 3.47.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/dx-scanner/v/3.47.0)
- [GitHub release](https://github.com/DXHeroes/dx-scanner/releases/tag/v3.47.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
